In [1]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import os
from pathlib import Path
import xgboost as xgb
import lightgbm as lgbm
import catboost
from sklearn.model_selection import StratifiedKFold, train_test_split
from sklearn.metrics import roc_auc_score
from IPython.display import display
from sklearn.preprocessing import LabelEncoder, OneHotEncoder, OrdinalEncoder
import optuna
from sklearn.preprocessing import StandardScaler
from scipy.linalg import norm

import tensorflow as tf
from tensorflow import keras
from tensorflow.keras import layers
import keras_tuner as kt

from category_encoders import LeaveOneOutEncoder

from IPython.display import display
from tqdm.notebook import tqdm

In [47]:
from warnings import filterwarnings
filterwarnings("ignore")

# Loading Data

In [2]:
BASE_PATH = Path("/kaggle/input/playground-series-s3e4/")

train = pd.read_csv(BASE_PATH / "train.csv")
test = pd.read_csv(BASE_PATH / "test.csv")

print(f"Training dataset has {len(train)} rows with " \
                    f"{sum(train.Class) / len(train) * 100 :.2}% fraud rows.")

Training dataset has 219129 rows with 0.21% fraud rows.


# Preprocessing

In [3]:
train["Hour"] = ((train.Time // 3600) % 24).astype("int")
test["Hour"] = ((test.Time // 3600) % 24).astype("int")

In [4]:
def time_of_day(hour):
    if hour >=0 and hour <= 6:
        return "night"
    elif hour <= 12:
        return "morning"
    elif hour <= 18:
        return "afternoon"
    else:
        return "evening"

In [27]:
train["Hour"] = ((train.Time // 3600) % 24).astype("category")
test["Hour"] = ((test.Time // 3600) % 24).astype("category")

In [28]:
train["TimeOfDay"] = train.Hour.map(time_of_day)
test["TimeOfDay"] = test.Hour.map(time_of_day)

In [29]:
train["Day"] = (((train['Time'] // (24 * 3600)) % 7) + 1).astype("category")
test["Day"] = (((test["Time"] // (24 * 3600)) % 7) + 1).astype("category")

In [30]:
train.head()

id  Time        V1        V2        V3        V4        V5        V6  \
0   0   0.0  2.074329 -0.129425 -1.137418  0.412846 -0.192638 -1.210144   
1   1   0.0  1.998827 -1.250891 -0.520969 -0.894539 -1.122528 -0.270866   
2   2   0.0  0.091535  1.004517 -0.223445 -0.435249  0.667548 -0.988351   
3   3   0.0  1.979649 -0.184949 -1.064206  0.120125 -0.215238 -0.648829   
4   4   0.0  1.025898 -0.171827  1.203717  1.243900 -0.636572  1.099074   

         V7        V8  ...       V24       V25       V26       V27       V28  \
0  0.110697 -0.263477  ... -0.110835 -0.291459  0.207733 -0.076576 -0.059577   
1 -1.029289  0.050198  ... -0.461928 -0.465491 -0.464655 -0.009413 -0.038238   
2  0.948146 -0.084789  ...  0.951233 -0.506919  0.085046  0.224458  0.087356   
3 -0.087826 -0.035367  ... -0.042939  0.000799 -0.096148 -0.057780 -0.073839   
4 -0.938651  0.569239  ... -0.262813  0.257834 -0.252829  0.108338  0.021051   

   Amount  Class  Hour  TimeOfDay  Day  
0    1.98      0   0.0      night  1.0  
1   84.00      0   0.0      night  1.0  
2    2.69      0   0.0      night  1.0  
3    1.00      0   0.0      night  1.0  
4    1.00      0   0.0      night  1.0  

[5 rows x 35 columns]

In [40]:
X = train.drop(columns=["id", "Class"])
y = train.Class
X_test = test.drop(columns=["id"])

# Preprocessing

## TODO:
1. One hot every categorical column but Hour. Use LeaveOneOut for Hour.

In [41]:
numerical_feats = list(set(X.columns) - set(["TimeOfDay", "Day", "Hour"]))

In [42]:
# encoding
# feats_to_one_hot = ["TimeOfDay", "Day"]
feats_to_loo = ["Hour", "TimeOfDay", "Day"]

## leave one out
loo = LeaveOneOutEncoder()
loo.fit(X[feats_to_loo], y)
X[feats_to_loo] = loo.transform(X[feats_to_loo])
X_test[feats_to_loo] = loo.transform(X_test[feats_to_loo])

## one_hotting
# X = pd.get_dummies(X)


# scaling
feats_to_scale = numerical_feats
sc = StandardScaler()
sc.fit(X[feats_to_scale])
X[feats_to_scale] = sc.transform(X[feats_to_scale])
X_test[feats_to_scale] = sc.transform(X_test[feats_to_scale])

In [43]:
X.shape

(219129, 33)

# Modelling

In [44]:
def cross_validate(X, y, model):
    N_FOLDS = 8
    all_scores = np.zeros(N_FOLDS)

    skf = StratifiedKFold(n_splits=N_FOLDS, shuffle=True, random_state=1337)

    for fold_id, (train_idx, val_idx) in enumerate(skf.split(X, y)):
        
        X_tr, X_val = X.iloc[train_idx], X.iloc[val_idx]
        y_tr, y_val = y.iloc[train_idx], y.iloc[val_idx]

               
        model.fit(X_tr, y_tr, 
                  eval_set=[(X_val, y_val)],
                  early_stopping_rounds=50,
                 verbose=False)
        
        y_pred = model.predict_proba(X_val)[:, 1]
        
        auc = roc_auc_score(y_val, y_pred)
        
        print(f"Fold {fold_id} \t auc: {auc}")
        
        all_scores[fold_id] = (auc)
    
    avg_auc = np.mean(all_scores)
    
    print(f"Avg AUC: {avg_auc}")

In [45]:
xgb_clf = xgb.XGBClassifier(tree_method="gpu_hist")

In [48]:
cross_validate(X, y, xgb_clf)

Fold 0 	 auc: 0.7885467185317059
Fold 1 	 auc: 0.8224168839737742
Fold 2 	 auc: 0.8101395054859795
Fold 3 	 auc: 0.8316166387226758
Fold 4 	 auc: 0.7715932401828614
Fold 5 	 auc: 0.8184812239704128
Fold 6 	 auc: 0.8208640396679128
Fold 7 	 auc: 0.8208553579711619
Avg AUC: 0.8105642010633105


# Tuning XGBoost 

In [50]:
def objective_xgb(trial, X, y):
    params = {
        'tree_method': "gpu_hist",
        'n_estimators': trial.suggest_int('n_estimators', 50, 1000),
        'max_depth': trial.suggest_int('max_depth', 2, 20),
        'learning_rate': trial.suggest_loguniform('learning_rate', 0.01, 1.0),
        'min_child_weight': trial.suggest_int('min_child_weight', 1, 10),
        'gamma': trial.suggest_loguniform('gamma', 0.00001, 10),
        'subsample': trial.suggest_float('subsample', 0.2, 1.0, step=0.05),
        'colsample_bytree': trial.suggest_float('colsample_bytree', 0.2, 1.0, step=0.05),
        'early_stopping_rounds': trial.suggest_int("early_stoppig_rounds", 40, 100)
    }
    # we're gonna train on the combined dataset but, we'll only calculate the validation score only on comp data

    N_FOLDS = 5
    all_scores = np.zeros(N_FOLDS)

    skf = StratifiedKFold(n_splits=N_FOLDS, shuffle=True, random_state=1337)

    for fold_id, (train_idx, val_idx) in enumerate(skf.split(X, y)):
        
        X_tr, X_val = X.iloc[train_idx], X.iloc[val_idx]
        y_tr, y_val = y.iloc[train_idx], y.iloc[val_idx]

        
        model = xgb.XGBClassifier(**params)
        model.fit(X_tr, y_tr, eval_set=[(X_val, y_val)], verbose=False)
        
        y_pred = model.predict_proba(X_val)[:, 1]
                
        auc = roc_auc_score(y_val, y_pred)        
        all_scores[fold_id] = auc
    
    avg_auc = np.mean(all_scores)
    
    print(f"Avg AUC: {avg_auc}")
    
    return avg_auc

In [52]:
study_xgb = optuna.create_study(study_name="xgboost_tuning", direction="maximize")
func = lambda trial: objective_xgb(trial, X, y)
study_xgb.optimize(func, n_trials=100)

[I 2023-01-30 16:00:18,798] A new study created in memory with name: xgboost_tuning
[I 2023-01-30 16:00:24,519] Trial 0 finished with value: 0.8212660962783176 and parameters: {'n_estimators': 570, 'max_depth': 5, 'learning_rate': 0.04993716478530016, 'min_child_weight': 10, 'gamma': 0.0019182532560568532, 'subsample': 0.30000000000000004, 'colsample_bytree': 0.9000000000000001, 'early_stoppig_rounds': 52}. Best is trial 0 with value: 0.8212660962783176.


Avg AUC: 0.8212660962783176


[I 2023-01-30 16:00:29,800] Trial 1 finished with value: 0.8142354854110987 and parameters: {'n_estimators': 460, 'max_depth': 8, 'learning_rate': 0.0843632780739971, 'min_child_weight': 2, 'gamma': 4.060401743495176e-05, 'subsample': 0.2, 'colsample_bytree': 0.45, 'early_stoppig_rounds': 59}. Best is trial 0 with value: 0.8212660962783176.


Avg AUC: 0.8142354854110987


[I 2023-01-30 16:00:32,319] Trial 2 finished with value: 0.7916385563988122 and parameters: {'n_estimators': 698, 'max_depth': 3, 'learning_rate': 0.35205335830670687, 'min_child_weight': 3, 'gamma': 0.0012584641321475962, 'subsample': 0.30000000000000004, 'colsample_bytree': 0.30000000000000004, 'early_stoppig_rounds': 47}. Best is trial 0 with value: 0.8212660962783176.


Avg AUC: 0.7916385563988122


[I 2023-01-30 16:00:37,613] Trial 3 finished with value: 0.8251634356769209 and parameters: {'n_estimators': 306, 'max_depth': 3, 'learning_rate': 0.04817015474648095, 'min_child_weight': 2, 'gamma': 4.306059217576984e-05, 'subsample': 0.25, 'colsample_bytree': 0.45, 'early_stoppig_rounds': 84}. Best is trial 3 with value: 0.8251634356769209.


Avg AUC: 0.8251634356769209


[I 2023-01-30 16:00:41,243] Trial 4 finished with value: 0.8119823195473405 and parameters: {'n_estimators': 356, 'max_depth': 15, 'learning_rate': 0.1565145592211108, 'min_child_weight': 3, 'gamma': 3.999690600642652, 'subsample': 0.8500000000000001, 'colsample_bytree': 0.35000000000000003, 'early_stoppig_rounds': 57}. Best is trial 3 with value: 0.8251634356769209.


Avg AUC: 0.8119823195473405


[I 2023-01-30 16:00:44,827] Trial 5 finished with value: 0.752153511698498 and parameters: {'n_estimators': 319, 'max_depth': 15, 'learning_rate': 0.6103846055388589, 'min_child_weight': 4, 'gamma': 0.13897555450493967, 'subsample': 0.2, 'colsample_bytree': 0.2, 'early_stoppig_rounds': 97}. Best is trial 3 with value: 0.8251634356769209.


Avg AUC: 0.752153511698498


[I 2023-01-30 16:00:54,083] Trial 6 finished with value: 0.8193753757077357 and parameters: {'n_estimators': 943, 'max_depth': 13, 'learning_rate': 0.03359655307442329, 'min_child_weight': 9, 'gamma': 0.009309947450091935, 'subsample': 0.4, 'colsample_bytree': 0.2, 'early_stoppig_rounds': 58}. Best is trial 3 with value: 0.8251634356769209.


Avg AUC: 0.8193753757077357


[I 2023-01-30 16:00:58,093] Trial 7 finished with value: 0.8156743648209975 and parameters: {'n_estimators': 336, 'max_depth': 5, 'learning_rate': 0.19575919237863998, 'min_child_weight': 9, 'gamma': 8.187013917353598, 'subsample': 0.8500000000000001, 'colsample_bytree': 0.6000000000000001, 'early_stoppig_rounds': 94}. Best is trial 3 with value: 0.8251634356769209.


Avg AUC: 0.8156743648209975


[I 2023-01-30 16:01:03,132] Trial 8 finished with value: 0.8190923502195933 and parameters: {'n_estimators': 218, 'max_depth': 13, 'learning_rate': 0.11892024557120431, 'min_child_weight': 5, 'gamma': 0.0028174532570543116, 'subsample': 0.5, 'colsample_bytree': 0.75, 'early_stoppig_rounds': 42}. Best is trial 3 with value: 0.8251634356769209.


Avg AUC: 0.8190923502195933


[I 2023-01-30 16:01:11,733] Trial 9 finished with value: 0.8216943499980738 and parameters: {'n_estimators': 544, 'max_depth': 7, 'learning_rate': 0.04063324005207817, 'min_child_weight': 6, 'gamma': 0.021570419441951358, 'subsample': 0.65, 'colsample_bytree': 0.25, 'early_stoppig_rounds': 60}. Best is trial 3 with value: 0.8251634356769209.


Avg AUC: 0.8216943499980738


[I 2023-01-30 16:01:15,627] Trial 10 finished with value: 0.6964484984800519 and parameters: {'n_estimators': 53, 'max_depth': 19, 'learning_rate': 0.010683031929910138, 'min_child_weight': 1, 'gamma': 1.4995953990585286e-05, 'subsample': 0.65, 'colsample_bytree': 0.55, 'early_stoppig_rounds': 82}. Best is trial 3 with value: 0.8251634356769209.


Avg AUC: 0.6964484984800519


[I 2023-01-30 16:01:23,873] Trial 11 finished with value: 0.819132538797333 and parameters: {'n_estimators': 686, 'max_depth': 2, 'learning_rate': 0.02992353439468006, 'min_child_weight': 7, 'gamma': 0.00018687285572881816, 'subsample': 0.65, 'colsample_bytree': 0.45, 'early_stoppig_rounds': 73}. Best is trial 3 with value: 0.8251634356769209.


Avg AUC: 0.819132538797333


[I 2023-01-30 16:01:30,520] Trial 12 finished with value: 0.8123764076268877 and parameters: {'n_estimators': 560, 'max_depth': 8, 'learning_rate': 0.06331888990116817, 'min_child_weight': 6, 'gamma': 0.06560878189492292, 'subsample': 0.55, 'colsample_bytree': 0.35000000000000003, 'early_stoppig_rounds': 72}. Best is trial 3 with value: 0.8251634356769209.


Avg AUC: 0.8123764076268877


[I 2023-01-30 16:01:35,690] Trial 13 finished with value: 0.7800564893785472 and parameters: {'n_estimators': 142, 'max_depth': 8, 'learning_rate': 0.020016246398710077, 'min_child_weight': 7, 'gamma': 0.00023613926702244388, 'subsample': 1.0, 'colsample_bytree': 0.65, 'early_stoppig_rounds': 84}. Best is trial 3 with value: 0.8251634356769209.


Avg AUC: 0.7800564893785472


[I 2023-01-30 16:01:42,128] Trial 14 finished with value: 0.8267400869709458 and parameters: {'n_estimators': 700, 'max_depth': 5, 'learning_rate': 0.04586226652220966, 'min_child_weight': 5, 'gamma': 1.5364511242556617e-05, 'subsample': 0.75, 'colsample_bytree': 0.45, 'early_stoppig_rounds': 66}. Best is trial 14 with value: 0.8267400869709458.


Avg AUC: 0.8267400869709458


[I 2023-01-30 16:01:47,291] Trial 15 finished with value: 0.8200962969517356 and parameters: {'n_estimators': 875, 'max_depth': 4, 'learning_rate': 0.07063838833207484, 'min_child_weight': 1, 'gamma': 1.2351825713689654e-05, 'subsample': 0.8, 'colsample_bytree': 0.45, 'early_stoppig_rounds': 82}. Best is trial 14 with value: 0.8267400869709458.


Avg AUC: 0.8200962969517356


[I 2023-01-30 16:02:06,796] Trial 16 finished with value: 0.8239496615300578 and parameters: {'n_estimators': 773, 'max_depth': 10, 'learning_rate': 0.020416860351863124, 'min_child_weight': 4, 'gamma': 7.463825043080805e-05, 'subsample': 0.75, 'colsample_bytree': 0.75, 'early_stoppig_rounds': 89}. Best is trial 14 with value: 0.8267400869709458.


Avg AUC: 0.8239496615300578


[I 2023-01-30 16:02:10,720] Trial 17 finished with value: 0.819290747236192 and parameters: {'n_estimators': 455, 'max_depth': 2, 'learning_rate': 0.09687021463027443, 'min_child_weight': 3, 'gamma': 0.0002631729777990541, 'subsample': 1.0, 'colsample_bytree': 0.5, 'early_stoppig_rounds': 68}. Best is trial 14 with value: 0.8267400869709458.


Avg AUC: 0.819290747236192


[I 2023-01-30 16:02:16,992] Trial 18 finished with value: 0.8246140423368198 and parameters: {'n_estimators': 815, 'max_depth': 6, 'learning_rate': 0.05908387387119985, 'min_child_weight': 5, 'gamma': 3.6559954420086874e-05, 'subsample': 0.45, 'colsample_bytree': 1.0, 'early_stoppig_rounds': 66}. Best is trial 14 with value: 0.8267400869709458.


Avg AUC: 0.8246140423368198


[I 2023-01-30 16:02:22,224] Trial 19 finished with value: 0.7987709880252094 and parameters: {'n_estimators': 659, 'max_depth': 10, 'learning_rate': 0.2227976871595905, 'min_child_weight': 2, 'gamma': 1.0473589702312514e-05, 'subsample': 0.35000000000000003, 'colsample_bytree': 0.65, 'early_stoppig_rounds': 77}. Best is trial 14 with value: 0.8267400869709458.


Avg AUC: 0.7987709880252094


[I 2023-01-30 16:02:26,449] Trial 20 finished with value: 0.8184636267933658 and parameters: {'n_estimators': 206, 'max_depth': 4, 'learning_rate': 0.11692739183645899, 'min_child_weight': 8, 'gamma': 0.0005771245150445741, 'subsample': 0.75, 'colsample_bytree': 0.35000000000000003, 'early_stoppig_rounds': 89}. Best is trial 14 with value: 0.8267400869709458.


Avg AUC: 0.8184636267933658


[I 2023-01-30 16:02:33,107] Trial 21 finished with value: 0.8237941838941095 and parameters: {'n_estimators': 799, 'max_depth': 6, 'learning_rate': 0.052233061323309654, 'min_child_weight': 5, 'gamma': 5.280231236233793e-05, 'subsample': 0.45, 'colsample_bytree': 1.0, 'early_stoppig_rounds': 66}. Best is trial 14 with value: 0.8267400869709458.


Avg AUC: 0.8237941838941095


[I 2023-01-30 16:02:39,601] Trial 22 finished with value: 0.8246113403067369 and parameters: {'n_estimators': 993, 'max_depth': 6, 'learning_rate': 0.05865854030981301, 'min_child_weight': 4, 'gamma': 5.4096312809580097e-05, 'subsample': 0.55, 'colsample_bytree': 0.8500000000000001, 'early_stoppig_rounds': 66}. Best is trial 14 with value: 0.8267400869709458.


Avg AUC: 0.8246113403067369


[I 2023-01-30 16:02:43,513] Trial 23 finished with value: 0.8204238651311477 and parameters: {'n_estimators': 828, 'max_depth': 2, 'learning_rate': 0.0790383124179043, 'min_child_weight': 5, 'gamma': 2.560821364372977e-05, 'subsample': 0.25, 'colsample_bytree': 1.0, 'early_stoppig_rounds': 72}. Best is trial 14 with value: 0.8267400869709458.


Avg AUC: 0.8204238651311477


[I 2023-01-30 16:02:50,109] Trial 24 finished with value: 0.8279928600699131 and parameters: {'n_estimators': 643, 'max_depth': 4, 'learning_rate': 0.042256268258819736, 'min_child_weight': 7, 'gamma': 6.895628336350295e-05, 'subsample': 0.4, 'colsample_bytree': 0.55, 'early_stoppig_rounds': 76}. Best is trial 24 with value: 0.8279928600699131.


Avg AUC: 0.8279928600699131


[I 2023-01-30 16:02:58,291] Trial 25 finished with value: 0.8211493881965659 and parameters: {'n_estimators': 647, 'max_depth': 4, 'learning_rate': 0.027767101585599946, 'min_child_weight': 7, 'gamma': 0.00014418735280330067, 'subsample': 0.35000000000000003, 'colsample_bytree': 0.55, 'early_stoppig_rounds': 79}. Best is trial 24 with value: 0.8279928600699131.


Avg AUC: 0.8211493881965659


[I 2023-01-30 16:03:04,752] Trial 26 finished with value: 0.82189964856973 and parameters: {'n_estimators': 446, 'max_depth': 3, 'learning_rate': 0.03661457089456185, 'min_child_weight': 8, 'gamma': 9.922257096439541e-05, 'subsample': 0.30000000000000004, 'colsample_bytree': 0.4, 'early_stoppig_rounds': 87}. Best is trial 24 with value: 0.8279928600699131.


Avg AUC: 0.82189964856973


[I 2023-01-30 16:03:27,866] Trial 27 finished with value: 0.8241954047576195 and parameters: {'n_estimators': 720, 'max_depth': 20, 'learning_rate': 0.04072952903356549, 'min_child_weight': 6, 'gamma': 0.00041784288303072696, 'subsample': 0.4, 'colsample_bytree': 0.55, 'early_stoppig_rounds': 77}. Best is trial 24 with value: 0.8279928600699131.


Avg AUC: 0.8241954047576195


[I 2023-01-30 16:03:44,127] Trial 28 finished with value: 0.8214263826384716 and parameters: {'n_estimators': 601, 'max_depth': 9, 'learning_rate': 0.021226916022446387, 'min_child_weight': 8, 'gamma': 2.1929025503183095e-05, 'subsample': 0.9000000000000001, 'colsample_bytree': 0.65, 'early_stoppig_rounds': 94}. Best is trial 24 with value: 0.8279928600699131.


Avg AUC: 0.8214263826384716


[I 2023-01-30 16:03:50,362] Trial 29 finished with value: 0.8254455210783143 and parameters: {'n_estimators': 482, 'max_depth': 5, 'learning_rate': 0.04712972589451733, 'min_child_weight': 10, 'gamma': 8.964844709261353e-05, 'subsample': 0.7, 'colsample_bytree': 0.75, 'early_stoppig_rounds': 49}. Best is trial 24 with value: 0.8279928600699131.


Avg AUC: 0.8254455210783143


[I 2023-01-30 16:03:56,742] Trial 30 finished with value: 0.8184587824954821 and parameters: {'n_estimators': 500, 'max_depth': 12, 'learning_rate': 0.08332059921534775, 'min_child_weight': 10, 'gamma': 1.0296102098005137e-05, 'subsample': 0.7, 'colsample_bytree': 0.75, 'early_stoppig_rounds': 48}. Best is trial 24 with value: 0.8279928600699131.


Avg AUC: 0.8184587824954821


[I 2023-01-30 16:04:02,846] Trial 31 finished with value: 0.8304400330014916 and parameters: {'n_estimators': 397, 'max_depth': 5, 'learning_rate': 0.0469468834731411, 'min_child_weight': 10, 'gamma': 9.952589990242697e-05, 'subsample': 0.6000000000000001, 'colsample_bytree': 0.8500000000000001, 'early_stoppig_rounds': 52}. Best is trial 31 with value: 0.8304400330014916.


Avg AUC: 0.8304400330014916


[I 2023-01-30 16:04:09,104] Trial 32 finished with value: 0.8310300896709879 and parameters: {'n_estimators': 414, 'max_depth': 5, 'learning_rate': 0.047688350515820065, 'min_child_weight': 10, 'gamma': 9.777458958828332e-05, 'subsample': 0.6000000000000001, 'colsample_bytree': 0.8500000000000001, 'early_stoppig_rounds': 52}. Best is trial 32 with value: 0.8310300896709879.


Avg AUC: 0.8310300896709879


[I 2023-01-30 16:04:14,580] Trial 33 finished with value: 0.8272851520930621 and parameters: {'n_estimators': 393, 'max_depth': 7, 'learning_rate': 0.06914113610340615, 'min_child_weight': 9, 'gamma': 0.0006374942484054125, 'subsample': 0.55, 'colsample_bytree': 0.8500000000000001, 'early_stoppig_rounds': 52}. Best is trial 32 with value: 0.8310300896709879.


Avg AUC: 0.8272851520930621


[I 2023-01-30 16:04:20,107] Trial 34 finished with value: 0.825004565463027 and parameters: {'n_estimators': 403, 'max_depth': 7, 'learning_rate': 0.07027118222315827, 'min_child_weight': 9, 'gamma': 0.0007925101789512711, 'subsample': 0.55, 'colsample_bytree': 0.9000000000000001, 'early_stoppig_rounds': 54}. Best is trial 32 with value: 0.8310300896709879.


Avg AUC: 0.825004565463027


[I 2023-01-30 16:04:24,365] Trial 35 finished with value: 0.8150273528624036 and parameters: {'n_estimators': 400, 'max_depth': 7, 'learning_rate': 0.10083869576991558, 'min_child_weight': 10, 'gamma': 0.0016109164672466747, 'subsample': 0.6000000000000001, 'colsample_bytree': 0.8500000000000001, 'early_stoppig_rounds': 43}. Best is trial 32 with value: 0.8310300896709879.


Avg AUC: 0.8150273528624036


[I 2023-01-30 16:04:31,876] Trial 36 finished with value: 0.8233073881945987 and parameters: {'n_estimators': 271, 'max_depth': 9, 'learning_rate': 0.05058655552038969, 'min_child_weight': 9, 'gamma': 0.0004764589378359416, 'subsample': 0.5, 'colsample_bytree': 0.8500000000000001, 'early_stoppig_rounds': 53}. Best is trial 32 with value: 0.8310300896709879.


Avg AUC: 0.8233073881945987


[I 2023-01-30 16:04:36,034] Trial 37 finished with value: 0.8216606755885032 and parameters: {'n_estimators': 394, 'max_depth': 3, 'learning_rate': 0.0778780957683008, 'min_child_weight': 10, 'gamma': 0.00016737777061841716, 'subsample': 0.6000000000000001, 'colsample_bytree': 0.9000000000000001, 'early_stoppig_rounds': 62}. Best is trial 32 with value: 0.8310300896709879.


Avg AUC: 0.8216606755885032


[I 2023-01-30 16:04:44,256] Trial 38 finished with value: 0.824029776068093 and parameters: {'n_estimators': 585, 'max_depth': 5, 'learning_rate': 0.02836301710340795, 'min_child_weight': 9, 'gamma': 3.6458558809702005e-05, 'subsample': 0.5, 'colsample_bytree': 0.8, 'early_stoppig_rounds': 55}. Best is trial 32 with value: 0.8310300896709879.


Avg AUC: 0.824029776068093


[I 2023-01-30 16:04:47,588] Trial 39 finished with value: 0.8280403396821676 and parameters: {'n_estimators': 276, 'max_depth': 4, 'learning_rate': 0.14355373086390658, 'min_child_weight': 8, 'gamma': 0.0009403848212009785, 'subsample': 0.45, 'colsample_bytree': 0.95, 'early_stoppig_rounds': 45}. Best is trial 32 with value: 0.8310300896709879.


Avg AUC: 0.8280403396821676


[I 2023-01-30 16:04:51,100] Trial 40 finished with value: 0.8254597445332831 and parameters: {'n_estimators': 262, 'max_depth': 4, 'learning_rate': 0.15704451446648132, 'min_child_weight': 8, 'gamma': 0.0032067371193494464, 'subsample': 0.4, 'colsample_bytree': 0.95, 'early_stoppig_rounds': 43}. Best is trial 32 with value: 0.8310300896709879.


Avg AUC: 0.8254597445332831


[I 2023-01-30 16:04:56,294] Trial 41 finished with value: 0.8243684340799767 and parameters: {'n_estimators': 373, 'max_depth': 6, 'learning_rate': 0.06305648794020315, 'min_child_weight': 9, 'gamma': 0.0010777436498418216, 'subsample': 0.45, 'colsample_bytree': 0.95, 'early_stoppig_rounds': 50}. Best is trial 32 with value: 0.8310300896709879.


Avg AUC: 0.8243684340799767


[I 2023-01-30 16:05:00,189] Trial 42 finished with value: 0.8231028701512841 and parameters: {'n_estimators': 305, 'max_depth': 3, 'learning_rate': 0.09490604321255398, 'min_child_weight': 10, 'gamma': 0.00036683049647530673, 'subsample': 0.55, 'colsample_bytree': 0.8, 'early_stoppig_rounds': 40}. Best is trial 32 with value: 0.8310300896709879.


Avg AUC: 0.8231028701512841


[I 2023-01-30 16:05:04,448] Trial 43 finished with value: 0.8150881461851321 and parameters: {'n_estimators': 440, 'max_depth': 7, 'learning_rate': 0.13246634559646797, 'min_child_weight': 8, 'gamma': 0.00010689032820748604, 'subsample': 0.6000000000000001, 'colsample_bytree': 0.7, 'early_stoppig_rounds': 45}. Best is trial 32 with value: 0.8310300896709879.


Avg AUC: 0.8150881461851321


[I 2023-01-30 16:05:08,591] Trial 44 finished with value: 0.8127608521009071 and parameters: {'n_estimators': 172, 'max_depth': 4, 'learning_rate': 0.03575797390983614, 'min_child_weight': 7, 'gamma': 0.0002533085755610447, 'subsample': 0.5, 'colsample_bytree': 0.8, 'early_stoppig_rounds': 51}. Best is trial 32 with value: 0.8310300896709879.


Avg AUC: 0.8127608521009071


[I 2023-01-30 16:05:11,796] Trial 45 finished with value: 0.800064481476085 and parameters: {'n_estimators': 516, 'max_depth': 5, 'learning_rate': 0.31088291934444495, 'min_child_weight': 9, 'gamma': 0.000920152213650767, 'subsample': 0.35000000000000003, 'colsample_bytree': 0.95, 'early_stoppig_rounds': 61}. Best is trial 32 with value: 0.8310300896709879.


Avg AUC: 0.800064481476085


[I 2023-01-30 16:05:19,330] Trial 46 finished with value: 0.8239597265267244 and parameters: {'n_estimators': 317, 'max_depth': 15, 'learning_rate': 0.053130423515025144, 'min_child_weight': 10, 'gamma': 0.0020442701283429413, 'subsample': 0.65, 'colsample_bytree': 0.9000000000000001, 'early_stoppig_rounds': 46}. Best is trial 32 with value: 0.8310300896709879.


Avg AUC: 0.8239597265267244


[I 2023-01-30 16:05:24,899] Trial 47 finished with value: 0.8219469817021349 and parameters: {'n_estimators': 349, 'max_depth': 3, 'learning_rate': 0.04381905601586443, 'min_child_weight': 8, 'gamma': 0.00015305519009313785, 'subsample': 0.45, 'colsample_bytree': 0.8500000000000001, 'early_stoppig_rounds': 57}. Best is trial 32 with value: 0.8310300896709879.


Avg AUC: 0.8219469817021349


[I 2023-01-30 16:05:29,232] Trial 48 finished with value: 0.8220048658304215 and parameters: {'n_estimators': 108, 'max_depth': 8, 'learning_rate': 0.06991340218549438, 'min_child_weight': 9, 'gamma': 6.258097095586396e-05, 'subsample': 0.5, 'colsample_bytree': 0.9000000000000001, 'early_stoppig_rounds': 52}. Best is trial 32 with value: 0.8310300896709879.


Avg AUC: 0.8220048658304215


[I 2023-01-30 16:05:34,554] Trial 49 finished with value: 0.8215604242562847 and parameters: {'n_estimators': 276, 'max_depth': 6, 'learning_rate': 0.09681894116244157, 'min_child_weight': 7, 'gamma': 0.0003259390655986304, 'subsample': 0.65, 'colsample_bytree': 0.7, 'early_stoppig_rounds': 63}. Best is trial 32 with value: 0.8310300896709879.


Avg AUC: 0.8215604242562847


[I 2023-01-30 16:05:36,927] Trial 50 finished with value: 0.7807930656564747 and parameters: {'n_estimators': 228, 'max_depth': 2, 'learning_rate': 0.9045021191314466, 'min_child_weight': 9, 'gamma': 2.911644303067532e-05, 'subsample': 0.55, 'colsample_bytree': 0.6000000000000001, 'early_stoppig_rounds': 56}. Best is trial 32 with value: 0.8310300896709879.


Avg AUC: 0.7807930656564747


[I 2023-01-30 16:05:43,520] Trial 51 finished with value: 0.8231877371757257 and parameters: {'n_estimators': 640, 'max_depth': 5, 'learning_rate': 0.04189228906832134, 'min_child_weight': 6, 'gamma': 8.595126675528249e-05, 'subsample': 0.7, 'colsample_bytree': 0.5, 'early_stoppig_rounds': 46}. Best is trial 32 with value: 0.8310300896709879.


Avg AUC: 0.8231877371757257


[I 2023-01-30 16:05:48,940] Trial 52 finished with value: 0.8244827051032303 and parameters: {'n_estimators': 702, 'max_depth': 4, 'learning_rate': 0.057522686218650386, 'min_child_weight': 10, 'gamma': 2.759276049757021e-05, 'subsample': 0.8500000000000001, 'colsample_bytree': 0.4, 'early_stoppig_rounds': 70}. Best is trial 32 with value: 0.8310300896709879.


Avg AUC: 0.8244827051032303


[I 2023-01-30 16:05:56,822] Trial 53 finished with value: 0.8254845001510102 and parameters: {'n_estimators': 740, 'max_depth': 5, 'learning_rate': 0.03230047388525645, 'min_child_weight': 8, 'gamma': 1.800308231653364e-05, 'subsample': 0.75, 'colsample_bytree': 0.5, 'early_stoppig_rounds': 58}. Best is trial 32 with value: 0.8310300896709879.


Avg AUC: 0.8254845001510102


[I 2023-01-30 16:06:07,544] Trial 54 finished with value: 0.813620705558699 and parameters: {'n_estimators': 529, 'max_depth': 17, 'learning_rate': 0.0485146734040034, 'min_child_weight': 7, 'gamma': 4.787571845614448e-05, 'subsample': 0.9000000000000001, 'colsample_bytree': 0.25, 'early_stoppig_rounds': 48}. Best is trial 32 with value: 0.8310300896709879.


Avg AUC: 0.813620705558699


[I 2023-01-30 16:06:13,592] Trial 55 finished with value: 0.8169388879580444 and parameters: {'n_estimators': 606, 'max_depth': 8, 'learning_rate': 0.06587389534309748, 'min_child_weight': 6, 'gamma': 0.00018864775538235908, 'subsample': 0.4, 'colsample_bytree': 0.4, 'early_stoppig_rounds': 59}. Best is trial 32 with value: 0.8310300896709879.


Avg AUC: 0.8169388879580444


[I 2023-01-30 16:06:21,215] Trial 56 finished with value: 0.8278333416826846 and parameters: {'n_estimators': 429, 'max_depth': 7, 'learning_rate': 0.03797700833988998, 'min_child_weight': 9, 'gamma': 1.8375847762944676e-05, 'subsample': 0.6000000000000001, 'colsample_bytree': 0.95, 'early_stoppig_rounds': 41}. Best is trial 32 with value: 0.8310300896709879.


Avg AUC: 0.8278333416826846


[I 2023-01-30 16:06:32,030] Trial 57 finished with value: 0.8250862086229214 and parameters: {'n_estimators': 434, 'max_depth': 7, 'learning_rate': 0.024665685728979852, 'min_child_weight': 9, 'gamma': 0.0006123561407068546, 'subsample': 0.6000000000000001, 'colsample_bytree': 0.95, 'early_stoppig_rounds': 40}. Best is trial 32 with value: 0.8310300896709879.


Avg AUC: 0.8250862086229214


[I 2023-01-30 16:06:41,055] Trial 58 finished with value: 0.8256581096382003 and parameters: {'n_estimators': 473, 'max_depth': 10, 'learning_rate': 0.03722161256018242, 'min_child_weight': 10, 'gamma': 5.092178966436189e-05, 'subsample': 0.55, 'colsample_bytree': 1.0, 'early_stoppig_rounds': 44}. Best is trial 32 with value: 0.8310300896709879.


Avg AUC: 0.8256581096382003


[I 2023-01-30 16:06:48,418] Trial 59 finished with value: 0.8228604373683238 and parameters: {'n_estimators': 375, 'max_depth': 6, 'learning_rate': 0.03432543480545376, 'min_child_weight': 9, 'gamma': 0.0001413906795639406, 'subsample': 0.45, 'colsample_bytree': 0.8, 'early_stoppig_rounds': 41}. Best is trial 32 with value: 0.8310300896709879.


Avg AUC: 0.8228604373683238


[I 2023-01-30 16:06:54,557] Trial 60 finished with value: 0.8177631481202357 and parameters: {'n_estimators': 549, 'max_depth': 9, 'learning_rate': 0.05769602181776898, 'min_child_weight': 8, 'gamma': 1.8619147351383288e-05, 'subsample': 0.2, 'colsample_bytree': 0.9000000000000001, 'early_stoppig_rounds': 100}. Best is trial 32 with value: 0.8310300896709879.


Avg AUC: 0.8177631481202357


[I 2023-01-30 16:07:00,389] Trial 61 finished with value: 0.8238475540888878 and parameters: {'n_estimators': 415, 'max_depth': 4, 'learning_rate': 0.044450831618038256, 'min_child_weight': 4, 'gamma': 1.6429099061326804e-05, 'subsample': 0.65, 'colsample_bytree': 0.95, 'early_stoppig_rounds': 77}. Best is trial 32 with value: 0.8310300896709879.


Avg AUC: 0.8238475540888878


[I 2023-01-30 16:07:08,823] Trial 62 finished with value: 0.8231968102945537 and parameters: {'n_estimators': 358, 'max_depth': 6, 'learning_rate': 0.040038689854592496, 'min_child_weight': 5, 'gamma': 3.8410383874130054e-05, 'subsample': 0.8, 'colsample_bytree': 0.45, 'early_stoppig_rounds': 75}. Best is trial 32 with value: 0.8310300896709879.


Avg AUC: 0.8231968102945537


[I 2023-01-30 16:07:16,066] Trial 63 finished with value: 0.823019444122363 and parameters: {'n_estimators': 322, 'max_depth': 5, 'learning_rate': 0.030714812262836594, 'min_child_weight': 10, 'gamma': 6.742080039925837e-05, 'subsample': 0.7, 'colsample_bytree': 0.6000000000000001, 'early_stoppig_rounds': 50}. Best is trial 32 with value: 0.8310300896709879.


Avg AUC: 0.823019444122363


[I 2023-01-30 16:07:20,865] Trial 64 finished with value: 0.8239539949794661 and parameters: {'n_estimators': 766, 'max_depth': 3, 'learning_rate': 0.05183733477485501, 'min_child_weight': 9, 'gamma': 1.1471985039211319e-05, 'subsample': 0.5, 'colsample_bytree': 1.0, 'early_stoppig_rounds': 47}. Best is trial 32 with value: 0.8310300896709879.


Avg AUC: 0.8239539949794661


[I 2023-01-30 16:07:25,933] Trial 65 finished with value: 0.819214598116943 and parameters: {'n_estimators': 854, 'max_depth': 7, 'learning_rate': 0.07624609853036578, 'min_child_weight': 7, 'gamma': 2.663916017330253e-05, 'subsample': 0.30000000000000004, 'colsample_bytree': 0.8500000000000001, 'early_stoppig_rounds': 64}. Best is trial 32 with value: 0.8310300896709879.


Avg AUC: 0.819214598116943


[I 2023-01-30 16:07:32,247] Trial 66 finished with value: 0.8262911110607499 and parameters: {'n_estimators': 499, 'max_depth': 4, 'learning_rate': 0.045078828007879615, 'min_child_weight': 8, 'gamma': 0.00010925697776437466, 'subsample': 0.35000000000000003, 'colsample_bytree': 0.55, 'early_stoppig_rounds': 81}. Best is trial 32 with value: 0.8310300896709879.


Avg AUC: 0.8262911110607499


[I 2023-01-30 16:07:37,021] Trial 67 finished with value: 0.8167132870176605 and parameters: {'n_estimators': 670, 'max_depth': 2, 'learning_rate': 0.0874740843268066, 'min_child_weight': 10, 'gamma': 0.00023122186948695437, 'subsample': 0.65, 'colsample_bytree': 0.30000000000000004, 'early_stoppig_rounds': 54}. Best is trial 32 with value: 0.8310300896709879.


Avg AUC: 0.8167132870176605


[I 2023-01-30 16:07:50,236] Trial 68 finished with value: 0.8263522956939339 and parameters: {'n_estimators': 618, 'max_depth': 11, 'learning_rate': 0.026232513473158192, 'min_child_weight': 4, 'gamma': 4.146604401192424e-05, 'subsample': 0.6000000000000001, 'colsample_bytree': 0.7, 'early_stoppig_rounds': 42}. Best is trial 32 with value: 0.8310300896709879.


Avg AUC: 0.8263522956939339


[I 2023-01-30 16:07:55,123] Trial 69 finished with value: 0.8178804277349719 and parameters: {'n_estimators': 287, 'max_depth': 5, 'learning_rate': 0.06785884352470861, 'min_child_weight': 9, 'gamma': 7.493749042372425e-05, 'subsample': 0.55, 'colsample_bytree': 0.5, 'early_stoppig_rounds': 52}. Best is trial 32 with value: 0.8310300896709879.


Avg AUC: 0.8178804277349719


[I 2023-01-30 16:08:01,615] Trial 70 finished with value: 0.8218909319723933 and parameters: {'n_estimators': 558, 'max_depth': 3, 'learning_rate': 0.03686519274943572, 'min_child_weight': 3, 'gamma': 1.4709494907368001e-05, 'subsample': 0.45, 'colsample_bytree': 0.8500000000000001, 'early_stoppig_rounds': 70}. Best is trial 32 with value: 0.8310300896709879.


Avg AUC: 0.8218909319723933


[I 2023-01-30 16:08:16,464] Trial 71 finished with value: 0.8264395464163569 and parameters: {'n_estimators': 642, 'max_depth': 12, 'learning_rate': 0.02448943787798628, 'min_child_weight': 4, 'gamma': 3.439207234295281e-05, 'subsample': 0.6000000000000001, 'colsample_bytree': 0.75, 'early_stoppig_rounds': 42}. Best is trial 32 with value: 0.8310300896709879.


Avg AUC: 0.8264395464163569


[I 2023-01-30 16:08:30,094] Trial 72 finished with value: 0.8209816219475196 and parameters: {'n_estimators': 719, 'max_depth': 13, 'learning_rate': 0.03213380822753016, 'min_child_weight': 3, 'gamma': 2.108241916651521e-05, 'subsample': 0.6000000000000001, 'colsample_bytree': 0.9000000000000001, 'early_stoppig_rounds': 45}. Best is trial 32 with value: 0.8310300896709879.


Avg AUC: 0.8209816219475196


[I 2023-01-30 16:08:45,028] Trial 73 finished with value: 0.8263855879480397 and parameters: {'n_estimators': 460, 'max_depth': 12, 'learning_rate': 0.023153042223937784, 'min_child_weight': 4, 'gamma': 3.506183438865695e-05, 'subsample': 0.55, 'colsample_bytree': 0.8, 'early_stoppig_rounds': 43}. Best is trial 32 with value: 0.8310300896709879.


Avg AUC: 0.8263855879480397


[I 2023-01-30 16:09:05,169] Trial 74 finished with value: 0.826469995130382 and parameters: {'n_estimators': 578, 'max_depth': 14, 'learning_rate': 0.01774317489501499, 'min_child_weight': 5, 'gamma': 6.234387376526966e-05, 'subsample': 0.65, 'colsample_bytree': 0.95, 'early_stoppig_rounds': 49}. Best is trial 32 with value: 0.8310300896709879.


Avg AUC: 0.826469995130382


[I 2023-01-30 16:09:28,369] Trial 75 finished with value: 0.8238689234063772 and parameters: {'n_estimators': 574, 'max_depth': 16, 'learning_rate': 0.01724766770258743, 'min_child_weight': 5, 'gamma': 0.0001330593982625925, 'subsample': 0.7, 'colsample_bytree': 0.95, 'early_stoppig_rounds': 49}. Best is trial 32 with value: 0.8310300896709879.


Avg AUC: 0.8238689234063772


[I 2023-01-30 16:09:37,856] Trial 76 finished with value: 0.8176112665645953 and parameters: {'n_estimators': 239, 'max_depth': 14, 'learning_rate': 0.01717522043839463, 'min_child_weight': 5, 'gamma': 0.00035220003248390385, 'subsample': 0.65, 'colsample_bytree': 1.0, 'early_stoppig_rounds': 74}. Best is trial 32 with value: 0.8310300896709879.


Avg AUC: 0.8176112665645953


[I 2023-01-30 16:09:43,916] Trial 77 finished with value: 0.8249396118509983 and parameters: {'n_estimators': 421, 'max_depth': 6, 'learning_rate': 0.060355564001428226, 'min_child_weight': 6, 'gamma': 6.476508410004172e-05, 'subsample': 0.5, 'colsample_bytree': 0.9000000000000001, 'early_stoppig_rounds': 54}. Best is trial 32 with value: 0.8310300896709879.


Avg AUC: 0.8249396118509983


[I 2023-01-30 16:09:50,727] Trial 78 finished with value: 0.8153636671966936 and parameters: {'n_estimators': 378, 'max_depth': 16, 'learning_rate': 0.05213063294408517, 'min_child_weight': 10, 'gamma': 0.00011084610855535211, 'subsample': 0.25, 'colsample_bytree': 0.95, 'early_stoppig_rounds': 51}. Best is trial 32 with value: 0.8310300896709879.


Avg AUC: 0.8153636671966936


[I 2023-01-30 16:09:59,401] Trial 79 finished with value: 0.8221357999828536 and parameters: {'n_estimators': 492, 'max_depth': 8, 'learning_rate': 0.03979215816295341, 'min_child_weight': 9, 'gamma': 0.00022233432927328103, 'subsample': 0.75, 'colsample_bytree': 0.45, 'early_stoppig_rounds': 47}. Best is trial 32 with value: 0.8310300896709879.


Avg AUC: 0.8221357999828536


[I 2023-01-30 16:10:04,768] Trial 80 finished with value: 0.8235644684393784 and parameters: {'n_estimators': 342, 'max_depth': 4, 'learning_rate': 0.047324685438400904, 'min_child_weight': 5, 'gamma': 1.0327512004955316e-05, 'subsample': 0.65, 'colsample_bytree': 0.65, 'early_stoppig_rounds': 56}. Best is trial 32 with value: 0.8310300896709879.


Avg AUC: 0.8235644684393784


[I 2023-01-30 16:10:17,971] Trial 81 finished with value: 0.8233416218433096 and parameters: {'n_estimators': 520, 'max_depth': 12, 'learning_rate': 0.0295629390354413, 'min_child_weight': 4, 'gamma': 2.3004503125759667e-05, 'subsample': 0.6000000000000001, 'colsample_bytree': 0.8500000000000001, 'early_stoppig_rounds': 45}. Best is trial 32 with value: 0.8310300896709879.


Avg AUC: 0.8233416218433096


[I 2023-01-30 16:10:32,123] Trial 82 finished with value: 0.8239561709899462 and parameters: {'n_estimators': 627, 'max_depth': 14, 'learning_rate': 0.02715546334139527, 'min_child_weight': 5, 'gamma': 4.8499403517804665e-05, 'subsample': 0.7, 'colsample_bytree': 0.8500000000000001, 'early_stoppig_rounds': 41}. Best is trial 32 with value: 0.8310300896709879.


Avg AUC: 0.8239561709899462


[I 2023-01-30 16:10:44,551] Trial 83 finished with value: 0.8242584342521951 and parameters: {'n_estimators': 585, 'max_depth': 13, 'learning_rate': 0.03322019992748196, 'min_child_weight': 2, 'gamma': 7.45128829480159e-05, 'subsample': 0.4, 'colsample_bytree': 0.75, 'early_stoppig_rounds': 49}. Best is trial 32 with value: 0.8310300896709879.


Avg AUC: 0.8242584342521951


[I 2023-01-30 16:10:53,069] Trial 84 finished with value: 0.824846876766063 and parameters: {'n_estimators': 685, 'max_depth': 11, 'learning_rate': 0.0402193233208021, 'min_child_weight': 6, 'gamma': 3.39728321047171e-05, 'subsample': 0.55, 'colsample_bytree': 0.75, 'early_stoppig_rounds': 44}. Best is trial 32 with value: 0.8310300896709879.


Avg AUC: 0.824846876766063


[I 2023-01-30 16:10:58,389] Trial 85 finished with value: 0.825455524082616 and parameters: {'n_estimators': 196, 'max_depth': 5, 'learning_rate': 0.04791849309422451, 'min_child_weight': 8, 'gamma': 0.000655544786320774, 'subsample': 0.8, 'colsample_bytree': 0.8, 'early_stoppig_rounds': 79}. Best is trial 32 with value: 0.8310300896709879.


Avg AUC: 0.825455524082616


[I 2023-01-30 16:11:03,938] Trial 86 finished with value: 0.8158308521216234 and parameters: {'n_estimators': 734, 'max_depth': 6, 'learning_rate': 0.07307583989752915, 'min_child_weight': 4, 'gamma': 1.7204376293651932e-05, 'subsample': 0.6000000000000001, 'colsample_bytree': 0.95, 'early_stoppig_rounds': 47}. Best is trial 32 with value: 0.8310300896709879.


Avg AUC: 0.8158308521216234


[I 2023-01-30 16:11:08,945] Trial 87 finished with value: 0.8207724510764104 and parameters: {'n_estimators': 650, 'max_depth': 4, 'learning_rate': 0.06073609167009717, 'min_child_weight': 10, 'gamma': 0.0002958558326734894, 'subsample': 0.5, 'colsample_bytree': 0.9000000000000001, 'early_stoppig_rounds': 53}. Best is trial 32 with value: 0.8310300896709879.


Avg AUC: 0.8207724510764104


[I 2023-01-30 16:11:18,001] Trial 88 finished with value: 0.8234580317339388 and parameters: {'n_estimators': 663, 'max_depth': 14, 'learning_rate': 0.053175269370512715, 'min_child_weight': 6, 'gamma': 0.0004666436363001517, 'subsample': 0.6000000000000001, 'colsample_bytree': 1.0, 'early_stoppig_rounds': 67}. Best is trial 32 with value: 0.8310300896709879.


Avg AUC: 0.8234580317339388


[I 2023-01-30 16:11:28,310] Trial 89 finished with value: 0.822429716509386 and parameters: {'n_estimators': 399, 'max_depth': 9, 'learning_rate': 0.0362587360952178, 'min_child_weight': 7, 'gamma': 9.466942237645003e-05, 'subsample': 0.65, 'colsample_bytree': 0.8, 'early_stoppig_rounds': 86}. Best is trial 32 with value: 0.8310300896709879.


Avg AUC: 0.822429716509386


[I 2023-01-30 16:11:42,956] Trial 90 finished with value: 0.823573160187455 and parameters: {'n_estimators': 771, 'max_depth': 11, 'learning_rate': 0.02252751495826187, 'min_child_weight': 8, 'gamma': 0.00015454793792982788, 'subsample': 0.75, 'colsample_bytree': 0.9000000000000001, 'early_stoppig_rounds': 42}. Best is trial 32 with value: 0.8310300896709879.


Avg AUC: 0.823573160187455


[I 2023-01-30 16:11:57,447] Trial 91 finished with value: 0.8213083185718265 and parameters: {'n_estimators': 474, 'max_depth': 12, 'learning_rate': 0.02422627457856493, 'min_child_weight': 4, 'gamma': 2.645514095657903e-05, 'subsample': 0.55, 'colsample_bytree': 0.8, 'early_stoppig_rounds': 44}. Best is trial 32 with value: 0.8310300896709879.


Avg AUC: 0.8213083185718265


[I 2023-01-30 16:12:13,611] Trial 92 finished with value: 0.8231461065561806 and parameters: {'n_estimators': 457, 'max_depth': 10, 'learning_rate': 0.019552360766194727, 'min_child_weight': 4, 'gamma': 3.7756396175837066e-05, 'subsample': 0.55, 'colsample_bytree': 0.8500000000000001, 'early_stoppig_rounds': 43}. Best is trial 32 with value: 0.8310300896709879.


Avg AUC: 0.8231461065561806


[I 2023-01-30 16:12:25,538] Trial 93 finished with value: 0.8258691899787778 and parameters: {'n_estimators': 437, 'max_depth': 12, 'learning_rate': 0.03077483217051446, 'min_child_weight': 5, 'gamma': 5.843717027636616e-05, 'subsample': 0.55, 'colsample_bytree': 0.7, 'early_stoppig_rounds': 40}. Best is trial 32 with value: 0.8310300896709879.


Avg AUC: 0.8258691899787778


[I 2023-01-30 16:12:33,620] Trial 94 finished with value: 0.8219088344257278 and parameters: {'n_estimators': 531, 'max_depth': 7, 'learning_rate': 0.04437021651885107, 'min_child_weight': 3, 'gamma': 1.4466826129545539e-05, 'subsample': 0.6000000000000001, 'colsample_bytree': 0.8500000000000001, 'early_stoppig_rounds': 46}. Best is trial 32 with value: 0.8310300896709879.


Avg AUC: 0.8219088344257278


[I 2023-01-30 16:12:45,787] Trial 95 finished with value: 0.8239574213000911 and parameters: {'n_estimators': 461, 'max_depth': 11, 'learning_rate': 0.02855640943633152, 'min_child_weight': 4, 'gamma': 3.149673064840892e-05, 'subsample': 0.45, 'colsample_bytree': 0.9000000000000001, 'early_stoppig_rounds': 50}. Best is trial 32 with value: 0.8310300896709879.


Avg AUC: 0.8239574213000911


[I 2023-01-30 16:13:05,499] Trial 96 finished with value: 0.823338600173269 and parameters: {'n_estimators': 691, 'max_depth': 13, 'learning_rate': 0.014572900278470203, 'min_child_weight': 9, 'gamma': 8.539029847997345e-05, 'subsample': 0.65, 'colsample_bytree': 0.75, 'early_stoppig_rounds': 43}. Best is trial 32 with value: 0.8310300896709879.


Avg AUC: 0.823338600173269


[I 2023-01-30 16:13:10,678] Trial 97 finished with value: 0.813600561440523 and parameters: {'n_estimators': 299, 'max_depth': 3, 'learning_rate': 0.02304384940536038, 'min_child_weight': 10, 'gamma': 4.457438235635496e-05, 'subsample': 0.5, 'colsample_bytree': 0.8, 'early_stoppig_rounds': 48}. Best is trial 32 with value: 0.8310300896709879.


Avg AUC: 0.813600561440523


[I 2023-01-30 16:13:25,106] Trial 98 finished with value: 0.8240176370736985 and parameters: {'n_estimators': 590, 'max_depth': 14, 'learning_rate': 0.026059708223839476, 'min_child_weight': 4, 'gamma': 0.00020270352014700206, 'subsample': 0.55, 'colsample_bytree': 0.55, 'early_stoppig_rounds': 52}. Best is trial 32 with value: 0.8310300896709879.


Avg AUC: 0.8240176370736985


[I 2023-01-30 16:13:32,104] Trial 99 finished with value: 0.8244837623738592 and parameters: {'n_estimators': 420, 'max_depth': 5, 'learning_rate': 0.038268803606100785, 'min_child_weight': 5, 'gamma': 2.38494934467478e-05, 'subsample': 0.7, 'colsample_bytree': 0.95, 'early_stoppig_rounds': 41}. Best is trial 32 with value: 0.8310300896709879.


Avg AUC: 0.8244837623738592


In [55]:
print(f"{'*'*15}XGBoost Tuning Results {'*'*15}\n\n"\
          f"BEST_VALUE: {study_xgb.best_value}\n\n"
          f"BEST_PARAMS: {study_xgb.best_params}"
     )

***************XGBoost Tuning Results ***************

BEST_VALUE: 0.8310300896709879

BEST_PARAMS: {'n_estimators': 414, 'max_depth': 5, 'learning_rate': 0.047688350515820065, 'min_child_weight': 10, 'gamma': 9.777458958828332e-05, 'subsample': 0.6000000000000001, 'colsample_bytree': 0.8500000000000001, 'early_stoppig_rounds': 52}


# Tuning LightGBM

In [83]:
from optuna.integration import LightGBMPruningCallback

def objective_lgbm(trial, X, y):
    param_grid = {
        "device_type": "gpu",
        "n_estimators": trial.suggest_int("n_estimators", 100, 1000),
        "num_rounds": trial.suggest_int("num_rounds", 100, 500),
        "learning_rate": trial.suggest_float("learning_rate", 0.0001, 1.0),
        "num_leaves": trial.suggest_int("num_leaves", 10, 200),
        "max_depth": trial.suggest_int("max_depth", 2, 20),
        "min_data_in_leaf": trial.suggest_int("min_data_in_leaf", 10, 200),
        "lambda_l1": trial.suggest_loguniform('lambda_l1', 0.01, 10),
        "lambda_l2": trial.suggest_loguniform('lambda_l2', 0.01, 10),
        "min_gain_to_split": trial.suggest_float("min_gain_to_split", 0.01, 15),
        "bagging_fraction":  trial.suggest_loguniform('bagging_fraction', 0.2, 1.0),
        "feature_fraction": trial.suggest_loguniform('feature_fraction', 0.2, 1.0),
        "early_stopping_rounds": trial.suggest_int("early_stopping_rounds", 50, 200)
    }

    
    
    N_FOLDS = 8
    all_scores = np.zeros(N_FOLDS)

    skf = StratifiedKFold(n_splits=N_FOLDS, shuffle=True, random_state=1337)

    for fold_id, (train_idx, val_idx) in enumerate(skf.split(X, y)):
        
        X_tr, X_val = X.iloc[train_idx], X.iloc[val_idx]
        y_tr, y_val = y.iloc[train_idx], y.iloc[val_idx]
            
        model = lgbm.LGBMClassifier(objective="binary", is_unbalance=True, **param_grid)
        model.fit(
            X_tr,
            y_tr,
            eval_set=[(X_val, y_val)],
            eval_metric="auc",
            verbose=-1,
        )
        y_preds = model.predict_proba(X_val)[:, 1]
        all_scores[fold_id] = roc_auc_score(y_val, y_preds)
    
    auc = np.mean(all_scores)
    print(f"AVG CV AUC: \t {auc}")
    return auc

In [84]:
study_lgbm = optuna.create_study(direction="maximize", study_name="LGBM Tuning")
func = lambda trial: objective_lgbm(trial, X, y)
study_lgbm.optimize(func, n_trials=100, show_progress_bar=True)

[I 2023-01-30 17:52:26,484] A new study created in memory with name: LGBM Tuning


  0%|          | 0/100 [00:00<?, ?it/s]

[LightGBM] [Warning] lambda_l2 is set=0.08959300755108116, reg_lambda=0.0 will be ignored. Current value: lambda_l2=0.08959300755108116
[LightGBM] [Warning] early_stopping_round is set=129, early_stopping_rounds=129 will be ignored. Current value: early_stopping_round=129
[LightGBM] [Warning] bagging_fraction is set=0.39584607013493944, subsample=1.0 will be ignored. Current value: bagging_fraction=0.39584607013493944
[LightGBM] [Warning] feature_fraction is set=0.8306731605955961, colsample_bytree=1.0 will be ignored. Current value: feature_fraction=0.8306731605955961
[LightGBM] [Warning] min_data_in_leaf is set=60, min_child_samples=20 will be ignored. Current value: min_data_in_leaf=60
[LightGBM] [Warning] num_iterations is set=462, num_rounds=462 will be ignored. Current value: num_iterations=462
[LightGBM] [Warning] min_gain_to_split is set=7.691706918724669, min_split_gain=0.0 will be ignored. Current value: min_gain_to_split=7.691706918724669
[LightGBM] [Warning] lambda_l1 is se

[LightGBM] [Fatal] Check failed: (best_split_info.left_count) > (0) at /usr/local/src/LightGBM/src/treelearner/serial_tree_learner.cpp, line 653 .



[W 2023-01-30 18:15:36,433] Trial 43 failed with parameters: {'n_estimators': 553, 'num_rounds': 148, 'learning_rate': 0.1516010227088807, 'num_leaves': 11, 'max_depth': 16, 'min_data_in_leaf': 26, 'lambda_l1': 0.4058699499893747, 'lambda_l2': 0.019698186826778995, 'min_gain_to_split': 6.617758481098144, 'bagging_fraction': 0.6134409783987361, 'feature_fraction': 0.46372469263998767, 'early_stopping_rounds': 184} because of the following error: LightGBMError('Check failed: (best_split_info.left_count) > (0) at /usr/local/src/LightGBM/src/treelearner/serial_tree_learner.cpp, line 653 .\n').
Traceback (most recent call last):
  File "/opt/conda/lib/python3.7/site-packages/optuna/study/_optimize.py", line 200, in _run_trial
    value_or_values = func(trial)
  File "/tmp/ipykernel_23/324840933.py", line 2, in <lambda>
    func = lambda trial: objective_lgbm(trial, X, y)
  File "/tmp/ipykernel_23/544366158.py", line 38, in objective_lgbm
    verbose=-1,
  File "/opt/conda/lib/python3.7/site

LightGBMError: Check failed: (best_split_info.left_count) > (0) at /usr/local/src/LightGBM/src/treelearner/serial_tree_learner.cpp, line 653 .


### best for now!
{'n_estimators': 625, 'num_rounds': 128, 'learning_rate': 0.0016209454169685287, 'num_leaves': 167, 'max_depth': 3, 'min_data_in_leaf': 106, 'lambda_l1': 0.12582407806544235, 'lambda_l2': 0.03324581293013046, 'min_gain_to_split': 10.560189522017625, 'bagging_fraction': 0.9946615300185752, 'feature_fraction': 0.4012736229918392, 'early_stopping_rounds': 140}.

### BEST v2 - again error stopped the optuna midway
Trial 40 finished with value: 0.820944282533944 and parameters: {'n_estimators': 536, 'num_rounds': 180, 'learning_rate': 0.000500503296430395, 'num_leaves': 15, 'max_depth': 14, 'min_data_in_leaf': 31, 'lambda_l1': 0.47264909156935453, 'lambda_l2': 0.024617360718187438, 'min_gain_to_split': 7.212881407911029, 'bagging_fraction': 0.6464435714593756, 'feature_fraction': 0.4710539166703898, 'early_stopping_rounds': 182}. Best is trial 40 with value: 0.820944282533944.

# Tuning CatBoost

In [80]:
def objective_cat(trial, X, y):
    param = {
        "task_type": "GPU"
        "n_estimators": trial.suggest_int("n_estimators", 100, 2000),
        "loss_function": trial.suggest_categorical("loss_function", ["CrossEntropy"]),
        "learning_rate": trial.suggest_loguniform("learning_rate", 1e-5, 1e0),
        "l2_leaf_reg": trial.suggest_loguniform("l2_leaf_reg", 1e-2, 1e0),
        "colsample_bylevel": trial.suggest_loguniform("colsample_bylevel", 0.01, 1.0),
        "depth": trial.suggest_int("depth", 1, 10),
        "boosting_type": trial.suggest_categorical("boosting_type", ["Ordered", "Plain"]),
        "bootstrap_type": trial.suggest_categorical("bootstrap_type", ["Bayesian", "Bernoulli", "MVS"]),
        "min_data_in_leaf": trial.suggest_int("min_data_in_leaf", 2, 20),
        "one_hot_max_size": trial.suggest_int("one_hot_max_size", 2, 20),
        "early_stopping_rounds": trial.suggest_int("early_stopping_rounds", 50, 200)
    }
    # Conditional Hyper-Parameters
    if param["bootstrap_type"] == "Bayesian":
        param["bagging_temperature"] = trial.suggest_float("bagging_temperature", 0, 10)
    elif param["bootstrap_type"] == "Bernoulli":
        param["subsample"] = trial.suggest_float("subsample", 0.1, 1)
    
    N_FOLDS = 8
    all_scores = np.zeros(N_FOLDS)

    skf = StratifiedKFold(n_splits=N_FOLDS, shuffle=True, random_state=1337)

    for fold_id, (train_idx, val_idx) in enumerate(skf.split(X, y)):
        
        X_tr, X_val = X.iloc[train_idx], X.iloc[val_idx]
        y_tr, y_val = y.iloc[train_idx], y.iloc[val_idx]

        cat_model = catboost.CatBoostClassifier(**param)
        cat_model.fit(X_tr, y_tr, eval_set=[(X_val, y_val)], verbose=False)
        
        y_preds = cat_model.predict_proba(X_val)[:, 1]
        all_scores[fold_id] = roc_auc_score(y_val, y_preds)
    
    auc = np.mean(all_scores)
    print(f"AVG CV AUC: \t {auc}")
    return auc

In [ ]:
study_cat = optuna.create_study(direction="maximize", study_name="CatBoost Tuning")
func = lambda trial: objective_cat(trial, X, y)
study_cat.optimize(func, n_trials=100, show_progress_bar=True)